### PR TITLE
Client, store, assessment, initApp

### LINKED ISSUE #1015: Home refactor
Depends on #1013

Home page non logged in user 

refactor store such as :
assessment: {
 assessment:Assessment,
countries:Array<Country>
regions, regionGroups.. .....
}

app is ready is assessment state is not undefined 

on app init action, call backend endpoint /api/init 

- implement country selection

### PR DESCRIPTION
**Pull request by sorja**
Resolves #1026 
Part of #1015

### CODE REVIEW

File: client/store/axiosErrorMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import axios, { AxiosStatic } from 'axios'
+
+import { Action, Dispatch, Middleware /* MiddlewareAPI */ } from 'redux'
+
+const createAxiosMiddleware =
+  (axios: AxiosStatic): Middleware =>
+  (/* { /!*dispatch*!/ }: MiddlewareAPI */) => {
+    axios.interceptors.response.use(null, (error: any) => {
+      // TODO: Handle dispatch, application error
+      // dispatch(applicationError(error))
```
Comment (sorja) [inline client/store/axiosErrorMiddleware.ts:10]: Commented off until error handling added

File: client/store/assessment/stateType.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Assessment } from '@core/meta/assessment'
+import { Country, Region } from '@core/country'
+import { RegionGroup } from '@core/meta/regionGroup'
+
+export interface AssessmentState {
+  assessment?: Assessment
+  countries?: Array<Country>
+  regions?: Array<Region>
+  regionGroups?: Array<RegionGroup>
```
Comment (sorja) [inline client/store/assessment/stateType.ts:9]: Should we use partial?

File: client/store/assessment/actions/initApp.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import axios from 'axios'
+import { ApiEndPoint } from '@common/api/endpoint'
+
+export const initApp = createAsyncThunk('app/init', async () => {
+  try {
+    return axios.get(ApiEndPoint.Init.one())
+  } catch (err) {
+    // if (err.response && err.response.status !== 401) {
+    //   dispatch(applicationError(err))
+    // }
```
Comment (sorja) [inline client/store/assessment/actions/initApp.ts:12]: Commented off until error handling added

File: client/store/assessment/stateType.ts (hunk 2 above)
Comment (minotogna) [reply client/store/assessment/stateType.ts:9]: what do you think about this structure:

```
export interface AssessmentState {
   assessment?: Assessment
   countries?: Array<Country>
   regionGroups?: Record<string, RegionGroup>
```
and RegionGroup has also an array of regions as property.
this way the /api/init could return such structure.

then your useRegionGroups hook will return an array of region groups sorted by order

File: client/store/assessment/actions/initApp.ts (hunk 3 above)
Comment (minotogna) [reply client/store/assessment/actions/initApp.ts:12]: we dont need the catch here because of the middleware?

File: client/store/index.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export { injectReducers, useAppSelector, useAppDispatch } from './store'
+export type { AppDispatch, ActionHandlers } from './store'
```
Comment (minotogna) [inline client/store/index.ts:2]: export also RootState

**Comment by minotogna** [comment]
does this really resolves #1015?

File: core/meta/regionGroup.ts [hunk 5]
```diff
@@ -1,5 +1,8 @@
+import { Region } from '@core/country'
```
Comment (minotogna) [inline core/meta/regionGroup.ts:1]: type Region must be added to core/meta. add it

**Review by minotogna — CHANGES_REQUESTED** [review]
can we also remove getRegions from repository ?

File: server/repository/assessment/getCountries.ts [hunk 6]
```diff
@@ -10,8 +10,8 @@ export const getCountries = async (props: { name: string }, client: BaseProtocol
   return client
     .many<Array<Country>>(
       `
-        select * from ${assessmentName}.assessment_country;
+        select country_iso from ${assessmentName}.country;
     `
     )
-    .then((data) => Objects.camelize(data))
+    .then((data) => Objects.camelize(data).map((d: Country) => d.countryIso))
```
Comment (minotogna) [inline server/repository/assessment/getCountries.ts:16]: this method should return an array of country, not country ISOs

File: core/meta/region.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { RegionCode } from '@core/country'
```
Comment (minotogna) [inline core/meta/region.ts:1]: we need to move all meta we use under core/meta, even region code

File: server/repository/assessment/getCountries.ts [hunk 8]
```diff
@@ -10,7 +10,7 @@ export const getCountries = async (props: { name: string }, client: BaseProtocol
   return client
     .many<Array<Country>>(
       `
-        select * from ${assessmentName}.assessment_country;
+          select c2.country_iso, config from ${assessmentName}.country c left join country c2 on c.country_iso = c2.country_iso;
```
Comment (minotogna) [inline server/repository/assessment/getCountries.ts:13]: let's not fetch all configs. just an array of country iso

**Review by minotogna — CHANGES_REQUESTED** [review]

File: server/repository/assessment/getCountries.ts (hunk 8 above)
Comment (minotogna) [reply server/repository/assessment/getCountries.ts:13]: can you avoid fetching config ?

**Review by minotogna — CHANGES_REQUESTED** [review]

File: core/country/index.ts [hunk 9]
```diff
@@ -3,5 +3,5 @@ export type { Country } from './country'
 export type { CountryIso } from './countryIso'
 export { Global } from './global'
 export type { Region } from './region'
-export { RegionCode } from './regionCode'
+export { RegionCode } from '../meta/regionCode'
 export type { RegionGroup } from './regionGroup'
```
Comment (minotogna) [inline core/country/index.ts:7]: all the objects we use, should be moved to meta

**Review by minotogna — CHANGES_REQUESTED** [review]

**Review by minotogna — APPROVED** [review]

**Pull request merged**